include/alpaka/event/EventGeneric.hpp

```diff
@@ -0,0 +1,508 @@
+                    EventGenericImpl(EventGenericImpl<TDev> &&) = delete;
+                    //-----------------------------------------------------------------------------
+                    auto operator=(EventGenericImpl<TDev> const &) -> EventGenericImpl<TDev> & = delete;
+                    //-----------------------------------------------------------------------------
+                    auto operator=(EventGenericImpl<TDev> &&) -> EventGenericImpl<TDev> & = delete;
+                    //-----------------------------------------------------------------------------
+                    ~EventGenericImpl() noexcept = default;
+
+                    //-----------------------------------------------------------------------------
+                    auto isReady() noexcept -> bool
+                    {
+                        return (m_LastReadyEnqueueCount == m_enqueueCount);
+                    }
+
+                    //-----------------------------------------------------------------------------
+                    auto wait(std::size_t const & enqueueCount, std::unique_lock<std::mutex>& lk) const noexcept -> void
+                    {
+                        ALPAKA_ASSERT(enqueueCount <= m_enqueueCount);
+
+                        while(enqueueCount > m_LastReadyEnqueueCount)
+                        {
+                            auto future = m_future;
+                            lk.unlock();
+                            future.get();
+                            lk.lock();
+                        }
+                    }
+
+                public:
+                    TDev const m_dev;                                //!< The device this event is bound to.
+
+                    std::mutex mutable m_mutex;                             //!< The mutex used to synchronize access to the event.
+                    std::shared_future<void> m_future;                      //!< The future signaling the event completion.
+                    std::size_t m_enqueueCount;                             //!< The number of times this event has been enqueued.
+                    std::size_t m_LastReadyEnqueueCount;                    //!< The time this event has been ready the last time.
+                                                                            //!< Ready means that the event was not waiting within a queue (not enqueued or already completed).
+                                                                            //!< If m_enqueueCount == m_LastReadyEnqueueCount, the event is currently not enqueued
+                };
+            }
+        }
+
+        //#############################################################################
+        //! The CPU device event.
+        template<
+            typename TDev>
+        class EventGeneric final : public concepts::Implements<wait::ConceptCurrentThreadWaitFor, EventGeneric<TDev>>
```

Comment:
As written in the other comment, I would really like to call this `EventCpuGeneric` and move it into a `cpu` subfolder.

Reply:
The only problem I see with the name `EventCpuGeneric` is that the upcoming OpenMP5 and OpenAcc backends are not only for CPU.